### DISCUSSION TITLE
use rewrite to eliminate logging overhead?

### DISCUSSION THREAD

**Discussion by jizhilong** [openrewrite/rewrite#2680] [Ideas] [answered]
In java applications, logging statement like `LOGGER.debug('a log with arg {}',  heavySerialization(object))` introduces much overhead, even when the logging emit level is set to INFO or above. 
   If transform the logging statement to the following one, the overhead should be eliminated:
```java
if (LOGGER.isDebugEnabled()) {
  LOGGER.debug('a log with arg {}',  heavySerialization(object));
}
```
   But the `LOGGER.isXXEnabled` guard is boilerplate, and developers may forget to add them.
   So, is it possible to develop a rewrite application to add the `if (LOGGER.isXXXEnabled)` guard statements automatically? So that developers can enable the rewriting with a maven plugin, to make sure every logging is guarded before feeded to javac.
   Alternatively, since there is no need to check the rewrite transformed code into version control system, maybe it is better to do the transform at AST level, work like a javac plugin or something similar?

**Comment by timtebeek** [comment, marked as answer]
Hi @jizhilong! Sorry it took a while to respond; any new discussions now trigger a message in Slack, so we don't miss any from now on.

I quite like your suggestion of (only) guarding logging statements that take a potentially expensive method invocation as an argument; it's similar to [this issue reported on the rewrite-logging-frameworks repository](https://github.com/openrewrite/rewrite-logging-frameworks/issues/81); you might want to subscribe to that issue. I think it strikes the right balance of only guarding potentially expensive method calls, such that boilerplate is kept to a minimum. If you're up for it we can even coach you to a recipe contribution, in which case it might be easiest to [coordinate via Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA).

As for your suggestion of changing the code only just before it gets compiled, and keep the code changes out of the code repository: we don't currently have any plans in that direction, but you're welcome to open a separate issue on this repository to track that separately from the logging guard suggestion.
